Add currying

lepoetemaudit opened the pull request:
This PR enables currying, where there aren't conflicts between multi-arity functions and curried applications in the manner discussed in #74.

Currying is currently enabled only for top level functions, and far references must also be imported before being curried. In future these limitations should be easy enough to remove, and the addition of anonymous functions to Alpaca will also help see more benefit from currying.

The curries themselves are generated as anonymous functions at the call site to minimise polluting modules and conflicts.

If a user attempts to curry in a way that the compiler cannot resolve unambiguously (say, passing two arguments to function `f` when both `f/3` and `f/5` are defined) then the compiler provides an error, listing the ambiguous curries.

This PR also contains some alterations to the parser to allow infix arguments to take expressions as arguments, not simply terms. This means that in combination with currying infix functions can be used to delineate expressions, reducing parenthesis usage (e.g, chaining |>, or using <| as in Elm or $ as in Haskell to avoid needing to nest expressions in parentheses).

Here is an example (adapted from curry.alp in the test_files) that demonstrates some of what this now enables.

```elm
module curry

-- Infix 'pipe forward' operator
let (|>) x f = f x

let add x y = 
  x + y

let filter_ pred l acc =
  match l with
      [] -> acc
    | hd :: tail ->
        match pred hd with
            true  -> filter_ pred tail hd :: acc
          | false -> filter_ pred tail acc

let filter pred l =
  filter_ pred l []

-- Currying this means we can create an '=' predicate function for use in filter/2
let eq a b =
  a == b

let filtered_list () =    
  [1, 2, 3] -- without currying, the pipe below would fail
  |> filter (eq 2)

let foo () =
  let adder = add 6 in
  (6 |> add 5 |> add 5, adder 20, filtered_list ())
```

Shout outs to @danabr, @OvermindDL1, @yurrriq, @saem, and of course @j14159 for their considerable amount of thought and participation on this issue.

j14159 commented:
@lepoetemaudit could you please include a section in either the `README.md` or `Tour.org` files to talk a little about how it works from the user perspective?

Otherwise I love this :)

danabr commented on src/alpaca_ast.hrl:
```diff
@@ -134,7 +134,7 @@
 
 -type alpaca_bits_type() :: int | float | binary | utf8.
-
+ 
```
Nitpick: Trailing whitespace.

danabr commented on src/alpaca_codegen.erl:
```diff
@@ -113,7 +113,7 @@ gen_fun(Env, #alpaca_fun_def{name={symbol, _, N}, versions=Vs}=Def) ->
                     FName = cerl:c_fname(list_to_atom(N), length(Args)),
                     A = [cerl:c_var(list_to_atom(X)) || {symbol, _, X} <- Args],
-                    {_, B} = gen_expr(Env, Body),
+                    {_, B} = gen_expr(Env, Body),                    
```
Nitpick: Trailing whitespace.

danabr commented on src/alpaca_codegen.erl:
```diff
@@ -290,17 +291,52 @@ gen_expr(Env, #alpaca_apply{expr={symbol, _Line, Name}, args=[{unit, _}]}) ->
+                _ ->  
+                    PossibleCurries = lists:filter(fun(X) -> X > DesiredArity end, AvailFuns),
+                    %% The typer ensures that we can curry unambiguously
+                    case PossibleCurries of
```
Not strictly necessary with a case here, as as you say the typer already guarantees it will be an unambiguous match. Hence: 
````
%% The typer ensures that we can curry unambiguously
[CurryArity] = lists:filter(fun(X) -> X > DesiredArity end, AvailFuns),
{cerl:c_fname(list_to_atom(Name), CurryArity), true, CurryArity}
````

danabr commented on src/alpaca_codegen.erl:
```diff
@@ -290,17 +291,52 @@ gen_expr(Env, #alpaca_apply{expr={symbol, _Line, Name}, args=[{unit, _}]}) ->
+        false -> Apply = cerl:c_apply(
+                    FName, 
+                    [A || {_, A} <- [gen_expr(Env, E) || E <- Args]]),
+    {               Env, Apply}
```
Not entirely sure what happened with the formatting here?

danabr commented on src/alpaca_codegen.erl:
```diff
@@ -290,17 +291,52 @@ gen_expr(Env, #alpaca_apply{expr={symbol, _Line, Name}, args=[{unit, _}]}) ->
+
+           gen_expr(Env2, Binding);
+
+        false -> Apply = cerl:c_apply(
```
This would probably be more readable if you pushed "Apply =" down to the next line.

danabr commented on src/alpaca_typer.erl:
```diff
@@ -1606,28 +1640,37 @@ typ_of(Env, Lvl, #alpaca_apply{line=L, expr=Expr, args=Args}) ->
+                                try
+                                    typ_apply(Env, Lvl, TypF, NextVar, Args, L)
+                                catch
+                                    error:{arity_error, _, _} = Err -> CurryFun(Err)
```
Since `{arity_error, _, _}` is module internal and used for control flow, it is more idiomatic to use `throw` rather than `error`.

danabr commented on src/alpaca_typer.erl:
```diff
@@ -1606,28 +1640,37 @@ typ_of(Env, Lvl, #alpaca_apply{line=L, expr=Expr, args=Args}) ->
     case typ_of(Env, Lvl, Expr) of
         {error, {bad_variable_name, _}} -> ForwardFun();
-        {error, _} = E -> E;
+        {error, _} = E -> E;            
```
Nitpick: Trailing whitespace.

lepoetemaudit commented:
@danabr - thanks for spotting all these, I will amend

j14159 commented:
@lepoetemaudit let me know when you're happy with this and I'll merge it and rebase mine again.

lepoetemaudit commented:
@j14159 sure, I just need to finish amending from @danabr's feedback and umm, figure out enough Emacs to export to markdown from `tour.org` :)

j14159 commented:
`M-x org-md-export-to-markdown` should do the trick if memory serves :)

lepoetemaudit commented:
That works, alas it gives me anchors like `<a id="orge61d414"></a>` instead of the nice ones like `<a href="#sec-1">`. Tried updating everything I could to no avail. I imagine this is down to my ignorance (or maybe it just knows I'm a Vim user and it's trolling me...)

j14159 commented:
Hrm, ok.  I don't think I modified the output of it when I last exported it.  Feel free to skip updating the markdown and I can do it in a separate PR after I rebase if you like.  Not a big deal.

lepoetemaudit commented:
I just manually synced the markdown file as it was only a small addition (and spotted a mistake I added in the tour as I did it), and I've fixed everything else. I enabled whitespace stripping in my editor and it's added a bit of noise to the diffs, so apologies for that.

j14159 commented:
Thanks again for this!
